File: LC/[0222] Count Complete Tree Nodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def countNodes(self, root: Optional[TreeNode]) -> int:
        if root is None: return 0

        def getHeight(rt):
            hh=0
            while rt:
                rt=rt.left
                hh+=1
            return hh


        def hasK(rt, k): #O(h) where h is logN
            if rt is None: return False
            
            path = []
            kk = k
            while kk>0:
                path.append(kk)
                kk//=2
            path.reverse()

            for nn in path[1:]:
                if nn%2==0:
                    rt = rt.left
                else:
                    rt = rt.right
                if rt is None: return False
            
            return True


        hh = getHeight(root)
        ll = pow(2, hh-1)
        rr = pow(2, hh)-1

        # binary search
        cnt = ll
        while ll <= rr:
            mm = ll + (rr-ll)//2
            if hasK(root, mm):
                cnt=mm
                ll=mm+1
            else:
                rr=mm-1

        return cnt



"""
## Problem statement
* given a complete tree -> count num of nodes with less than O(N) solution
## Analysis
* N=5*10^4
* value of the node is red herring

## strategy
* create a method hasK

* binary search the max

"""
        

File: LC/[0222] Count Complete Tree Nodes.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private boolean hasK(TreeNode node, int k){
        if(node == null) return false;
        //build the path
        ArrayList<Integer> path = new ArrayList<>();
        while(k>0){
            path.add(k);
            k /= 2;
        }

        //reverse the arrayList
        Collections.reverse(path);

        for(int ii=1; ii<path.size(); ++ii){
            int val = path.get(ii);
            if(val%2 != 0){
                node=node.right;
            }else{
                node=node.left;
            }
            if(node==null) return false;
        }

        //walk the path
        return true;
    }
    private int getHeight(TreeNode node){
        if(node == null) return 0;//no node -> height 0
        int hh=0;
        while(node != null){
            ++hh; //height starts from 1
            node=node.left;//next level -> keep moving left child
        }
        return hh;
    }
    public int countNodes(TreeNode root) {
        if(root == null) return 0; //no node at all

        //binary search
        int hh = getHeight(root);
        
        // -- define the range
        int ll = 1<<(hh-1);
        int rr = (1<<hh)-1;
        int ans=ll;
        while(ll<=rr){//binary search template
            int mm = ll+(rr-ll)/2;
            if(hasK(root, mm)){
                ans=mm;
                ll=mm+1;
            }else{
                rr=mm-1;
            }
        }

        return ans;
    }
    
}